File: src/main/java/amod/aido/demo/DispatchingLogic.java
/* amod - Copyright (c) 2018, ETH Zurich, Institute for Dynamic Systems and Control */
package amod.aido.demo;

import java.util.HashSet;
import java.util.Set;
import java.util.SortedMap;
import java.util.TreeMap;

import ch.ethz.idsc.amodeus.dispatcher.core.RoboTaxiStatus;
import ch.ethz.idsc.tensor.RealScalar;
import ch.ethz.idsc.tensor.Scalar;
import ch.ethz.idsc.tensor.Tensor;
import ch.ethz.idsc.tensor.Tensors;
import ch.ethz.idsc.tensor.pdf.Distribution;
import ch.ethz.idsc.tensor.pdf.RandomVariate;
import ch.ethz.idsc.tensor.pdf.UniformDistribution;
import ch.ethz.idsc.tensor.sca.Round;

/** dispatching logic in the AidoGuest demo to compute dispatching instructions
 * that are forwarded to the AidoHost */
/* package */ class DispatchingLogic {

    private final Set<Scalar> matchedReq = new HashSet<>();
    private final Set<Scalar> matchedTax = new HashSet<>();
    private final Distribution dist_lng;
    private final Distribution dist_lat;

    /** @param bottomLeft {lngMin, latMin}
     * @param topRight {lngMax, latMax} */
    public DispatchingLogic(Tensor bottomLeft, Tensor topRight) {
        Scalar lngMin = bottomLeft.Get(0);
        Scalar lngMax = topRight.Get(0);
        Scalar latMin = bottomLeft.Get(1);
        Scalar latMax = topRight.Get(1);

        System.out.println("minimum longitude in network: " + lngMin);
        System.out.println("maximum longitude in network: " + lngMax);
        System.out.println("minimum latitude  in network: " + latMin);
        System.out.println("maximum latitude  in network: " + latMax);

        /** Example:
         * minimum longitude in network: -71.38020297181387
         * maximum longitude in network: -70.44406349551404
         * minimum latitude in network: -33.869660953686626
         * maximum latitude in network: -33.0303523690584 */

        dist_lng = UniformDistribution.of(lngMin, lngMax);
        dist_lat = UniformDistribution.of(latMin, latMax);
    }

    public Tensor of(Tensor status) {
        Tensor pickup = Tensors.empty();
        Tensor rebalance = Tensors.empty();

        Scalar time = status.Get(0);
        if (Round.toMultipleOf(RealScalar.of(60)).apply(time).equals(time)) { // every minute
            int index = 0;

            /** sort requests according to submission time */
            SortedMap<Scalar, Tensor> requests = new TreeMap<>();
            for (Tensor request : status.get(2)) {
                requests.put(request.Get(1), request);
            }

            /** for each unassigned request, add a taxi in STAY mode */
            for (Tensor request : requests.values()) {
                if (!matchedReq.contains(request.Get(0))) {
                    while (index < status.get(1).length()) {
                        Tensor roboTaxi = status.get(1, index);
                        if (RoboTaxiStatus.valueOf(roboTaxi.Get(2).toString())//
                                .equals(RoboTaxiStatus.STAY)) {
                            pickup.append(Tensors.of(roboTaxi.Get(0), request.Get(0)));
                            matchedReq.add(request.Get(0));
                            matchedTax.add(roboTaxi.Get(0));
                            ++index;
                            break;
                        }
                        ++index;
                    }
                }
            }

            /** rebalance 1 of the remaining and unmatched STAY taxis */
            for (int i = 0; i < status.get(1).length(); ++i) {
                Tensor roboTaxi = status.get(1, i);
                if (RoboTaxiStatus.valueOf(roboTaxi.Get(2).toString())//
                        .equals(RoboTaxiStatus.STAY)) {
                    if (!matchedTax.contains(roboTaxi.Get(0))) {
                        Tensor rebalanceLocation = getRandomRebalanceLocation();
                        rebalance.append(Tensors.of(roboTaxi.Get(0), rebalanceLocation));
                        break;
                    }
                }
            }
        }
        return Tensors.of(pickup, rebalance);
    }

    private Tensor getRandomRebalanceLocation() {
        /** ATTENTION: AMoDeus internally uses the convention
         * (longitude, latitude) for a WGS:84 pair,
         * not the other way around as in some other cases. */
        return Tensors.of( //
                RandomVariate.of(dist_lng), //
                RandomVariate.of(dist_lat));
    }
}

File: src/main/python/DispatchingLogic.py
import random
from utils.RoboTaxiStatus import RoboTaxiStatus


class DispatchingLogic:
    """
    dispatching logic in the AidoGuest demo to compute dispatching instructions that are forwarded to the AidoHost
    """
    def __init__(self, bottomLeft, topRight):
        """
        :param bottomLeft: {lngMin, latMin}
        :param topRight: {lngMax, latMax}
        """
        self.lngMin = bottomLeft[0]
        self.lngMax = topRight[0]
        self.latMin = bottomLeft[1]
        self.latMax = topRight[1]

        print("minimum longitude in network: ", self.lngMin)
        print("maximum longitude in network: ", self.lngMax)
        print("minimum latitude  in network: ", self.latMin)
        print("maximum latitude  in network: ", self.latMax)

        # Example:
        # minimum longitude in network: -71.38020297181387
        # maximum longitude in network: -70.44406349551404
        # minimum latitude in network: -33.869660953686626
        # maximum latitude in network: -33.0303523690584

        self.matchedReq = set()
        self.matchedTax = set()

    def of(self, status):
        assert isinstance(status, list)
        pickup = []
        rebalance = []

        time = status[0]
        if time % 60 == 0:  # every minute
            index = 0

            # sort requests according to submission time
            requests = sorted(status[2].copy(), key=lambda request: request[1])

            # for each unassigned request, add a taxi in STAY mode
            for request in requests:
                if request[0] not in self.matchedReq:
                    while index < len(status[1]):
                        roboTaxi = status[1][index]
                        if roboTaxi[2] is RoboTaxiStatus.STAY:
                            pickup.append([roboTaxi[0], request[0]])
                            self.matchedReq.add(request[0])
                            self.matchedTax.add(roboTaxi[0])
                            index += 1
                            break
                        index += 1

            # rebalance 1 of the remaining and unmatched STAY taxis
            for roboTaxi in status[1]:
                if roboTaxi[2] is RoboTaxiStatus.STAY and roboTaxi[0] not in self.matchedTax:
                    rebalanceLocation = self.getRandomRebalanceLocation()
                    rebalance.append([roboTaxi[0], rebalanceLocation])
                    break

        return [pickup, rebalance]

    def getRandomRebalanceLocation(self):
        """
        ATTENTION: AMoDeus internally uses the convention (longitude, latitude) for a WGS:84 pair, not the other way
        around as in some other cases.
        """
        return [random.uniform(self.lngMin, self.lngMax),
                random.uniform(self.latMin, self.latMax)]
